### PR TITLE
SQMAGOPC-434: New API: Cancel recurring order

### PR DESCRIPTION
**Pull request by bartoszkaluzny-solteq**
- Creates endpoint to cancel subscription.
PS. Please check FI translations

### CODE REVIEW

File: Model/SubscriptionManagement.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+     * @return string
+     */
+    public function cancelSubscription($subscriptionId)
+    {
+        try {
+            $subscription = $this->subscriptionRepository->get((int)$subscriptionId);
+            $orderIds = $this->subscriptionLinkRepository->getOrderIdsBySubscriptionId((int)$subscriptionId);
+            $params = $this->request->getParams();
+
+            $customer = $this->customerRepository->get($params['username']);
```
Comment (kasianmarszalek) [inline Model/SubscriptionManagement.php:98]: You can get the customer via first checking if the customer is logged in: $this->customerSession->isLoggedIn() and then  $customer = $this->customerSession->getCustomer();

**Review by kasianmarszalek — CHANGES_REQUESTED** [review]
Change fetching customer to use the customerSession instead

File: etc/webapi.xml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0"?>
+<routes xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Webapi:etc/webapi.xsd">
+
+    <!-- Subscriptions Service -->
+    <route url="/V1/paytrail/subscription/cancel/:subscription_id" method="GET">
+        <service class="Paytrail\PaymentService\Api\SubscriptionManagementInterface" method="cancelSubscription" />
+        <resources>
+            <resource ref="anonymous"/>
```
Comment (tommi-martin) [inline etc/webapi.xml:9]: Incorrect resource declaration. This resource needs to be restricted to customer only. Right now it's open to everyone. 

See vendor/magento/module-customer/etc/webapi.xml for example:

```
    <route url="/V1/customers/me/password" method="PUT">
        <service class="Magento\Customer\Api\AccountManagementInterface" method="changePasswordById"/>
        <resources>
            <resource ref="self"/>
        </resources>
        <data>
            <parameter name="customerId" force="true">%customer_id%</parameter>
        </data>
    </route>
```

**Review by tommi-martin — CHANGES_REQUESTED** [review]

File: i18n/fi_FI.csv [hunk 3]
```diff
@@ -110,3 +110,5 @@ Hello,Hei
 "Select Card","Valitse kortti"
 "Unable to change card","Korttia ei voitu vaihtaa"
 "Card for subscription changed successfully","Kortin vaihto toistuvaan tilaukseen onnistui"
+"Subscription couldn't be canceled","Tilausta ei voitu peruuttaa"
+"Subscription has been canceled correctly","Tilaus on peruutettu oikein"
```
Comment (tommi-martin) [inline i18n/fi_FI.csv:114]: While the translation is alright it is not something native speaker would say.
You could shorten it to "Tilaus on peruutettu" 
The "oikein" (translated 'correctly') is a bit redundant in finnish.

Comment (bartoszkaluzny-solteq) [reply i18n/fi_FI.csv:114]: Done

File: Model/SubscriptionManagement.php (hunk 1 above)
Comment (bartoszkaluzny-solteq) [reply Model/SubscriptionManagement.php]: Done

**Review by kasianmarszalek — APPROVED** [review]

File: Model/SubscriptionManagement.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+     */
+    public function cancelSubscription($subscriptionId)
+    {
+        try {
+            $subscription = $this->subscriptionRepository->get((int)$subscriptionId);
+            $orderIds = $this->subscriptionLinkRepository->getOrderIdsBySubscriptionId((int)$subscriptionId);
+            $customer = $this->customerSession->getCustomer();
+
+            foreach ($orderIds as $orderId) {
+                $order = $this->orderRepository->get($orderId);
```
Comment (tommi-martin) [inline Model/SubscriptionManagement.php:94]: Performance warning repository load inside a loop. Prefer the usage of collection or getList methods over repository get when you are loading multiple entities.

File: Model/SubscriptionManagement.php [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        OrderManagementInterface $orderManagementInterface,
+        RequestInterface $request,
+        LoggerInterface $logger
+    ) {
+        $this->customerSession = $customerSession;
+        $this->subscriptionRepository = $subscriptionRepository;
+        $this->subscriptionLinkRepository = $subscriptionLinkRepository;
+        $this->orderRepository = $orderRepository;
+        $this->orderManagementInterface = $orderManagementInterface;
+        $this->request = $request;
```
Comment (tommi-martin) [inline Model/SubscriptionManagement.php:77]: Unused dependency: request

**Review by tommi-martin — APPROVED** [review]

**Pull request merged**